src/ordered_hash_set.py

```diff
@@ -39,6 +44,8 @@ def remove(self, item):
 
         :param item: (object), Removed item
         """
+        ordered_items = OrderedDict(self._items)
```

Comment:
self._item is a dictionary and it does not remember insertion order (for Python 3.5). We are creating an OrderedDict instance from an unordered data structure. Therefore, ordered_items is not ordered as well.

Also, we are creating an OrderedDict instance from a dictionary every time when adding and removing an element. We should also check that whether this implementation is efficient.

Reply:
Currently we are using doubly linked list to provide insertion ordering. Maybe we can initialize `self._items` as OrderedDict to provide this functionality instead.

Reply:
Btw, I like your approach. It makes the code cleaner.

Reply:
Actually that was my main question. Why is self._items not an OrderedDict. This would make much more sense and simply the whole implementation

Reply:
Why do you need a double linked list ?

Reply:
I agree that your approach makes sense. I preferred this way for the sake of more control. 

For instance, let's consider the  [drain](https://github.com/buyalsky/ordered-hash-set/blob/e6c09d81849cbb832099792b2411861dd01c687f/src/ordered_hash_set.py#L76) method. If you set the reverse argument as True it yields the elements from end to start. If we use OrderecDict, the drain function will look like this:

```python3
def drain(self, reverse=False):
    if reverse:
        reverse_list = reversed(self_items.keys())
        for key in reverse_list:
            yield self.items.pop(key)
    
```

The code looks simpler but there is a problem. The keys() method creates a list-like object, after that we are reversing it. Imagine a situation where the self._items has too many entries in it. We are creating a big data structure (reverse_list) and may cause memory inefficiency. 

Maybe there is a better way to overcome this. Also, this is an edge case so we can even ignore it.

Currently, we are not creating another data structure. Instead, we are iterating using the linked list. If the reversed argument passed as True. We are starting from self._last (which points to the last item), popping the item and yielding it. 
```python3
def drain(self, reverse=False):
    if reverse:
        while self._last:
            item = self._items.pop(self._last)
            yield item[0]
            self._last = item[1][0]
```

OrderedDict probably uses a linked or regular list in order to guarantee insertion order. 
However, if OrderedDict is implemented in C, it might make sense to implement for making OrderedSet faster.

Reply:
If these are items from a set. Why do you use a dict in the first place and not a simple list ? By making each element on the list unique, you can retrieve the element from the list easily given an index.

And you don't need a double linked list in that case as you can use index(), slicing, reversing of that list if necessary.

Or am I missing something ?

Reply:
From my understanding:
You're only playing with sort, pop, drain, etc, why not just use a list?

Once you have your item sorted, removed added or whatsoever,  you can update a dict (ordered or not it is not important anymore) where you need to have a `key`/`value` match and update or remove that entry based on that particular key.

Reply:
If we use list then membership checking would have O(n) complexity instead of O(1). Below I shared a small script so you can make a benchmark on your own. 

```python3
import timeit
# create a set that contains items from 0 to 9999999
s = {i for i in range(10000000)}
# create a list that contains items from 0 to 9999999
l = [i for i in range(10000000)]

# check whether -1 is in list and set
# we use timeit module to measure the execution times
list_execution_time = timeit.timeit('-1 in l', 'from __main__ import l', number=100)
set_execution_time = timeit.timeit('-1 in s', 'from __main__ import s', number=100)

print(f"list execution time: {list_execution_time}")
print(f"set execution time: {set_execution_time}")
print(f"set is {list_execution_time/set_execution_time} faster")
```

You can check this [link](https://stackoverflow.com/a/8929320) for more info about list vs set.

Reply:
Indeed you're right on this point, a list would have a O(n) complexity.

How about using this: https://pypi.org/project/ordered-set/ ?

Reply:
Yes, just like mine, there are another 3rd party ordered set implementations. I just wanted to implement this on my own. 

But, I am also planning to add more features. For instance providing thread-safe and asyncio-aware versions. 

I name this module as `ordered-hash-set` since `ordered-set` is already occupied on PyPI.

Reply:
Maybe a good implementation would be to have a class OrderedHashSet that inherits from OrderedDict and reimplement it accordingly.

Reply:
Sorry for the delayed response. Feel free to submit a PR if you manage to implement your idea. 

I have to close this since the other contributor's PR was merged. Btw, I'm planning to create another issue for adding benchmarks to the repo. You can let me know if you are interested.